src/Monarch/VirtualDom/VirtualDomTree.purs

```diff
@@ -22,88 +24,104 @@ where
 import Prelude
 import Undefined
 import Type.Row                                            as Row
+import Type.RowList (kind RowList, class RowToList)
+import Type.RowList as RowList
 import Monarch.VirtualDom.Facts
 import Monarch.VirtualDom.Facts.Hooks
 import Monarch.VirtualDom.NS
 import Monarch.VirtualDom.NS as NS
 import Monarch.Type.Row                                    as Row
+import Monarch.Type.Maybe
 
-foreign import data VirtualDomTree :: # Type -> Type -> Type
+foreign import data VirtualDomTree :: Maybe -> # Type -> Type -> Type
 
-foreign import fmapVirtualDomTree :: forall slots a b. (a -> b) -> VirtualDomTree slots a -> VirtualDomTree slots b
+foreign import fmapVirtualDomTree :: forall key slots a b. (a -> b) -> VirtualDomTree key slots a -> VirtualDomTree key slots b
 
-instance functorVirtualDomTree :: Functor (VirtualDomTree slots) where
+instance functorVirtualDomTree :: Functor (VirtualDomTree key slots) where
   map = fmapVirtualDomTree
 
 -- Hyperscript
 
 type Node (mkProperties :: # Type -> # Type)
           (mkOutputs    :: Type -> # Type -> # Type)
           (mkAttributes :: # Type -> # Type)
-  = forall r _r attributes hooks slots message
-  . Row.Union r _r (Facts mkProperties (mkOutputs message) attributes hooks)
+  = forall r _r k key _key attributes hooks slots message
+  . Row.Union r _r (Facts mkProperties (mkOutputs message) attributes hooks k)
  => Row.OptionalRecordCons r "attrs" (mkAttributes ()) attributes
  => Row.OptionalRecordCons r "hooks" (Hooks message) hooks
+ => ExtractKeyType r key
  => { | r }
- -> Array (VirtualDomTree slots message)
- -> VirtualDomTree slots message
+ -> Array (VirtualDomTree _key slots message)
```

Comment:
As a convention we prefix trash type variables with `_`.
For meaningful type variables I suggest to suffix it with a `'`.